File: julia/2022/day_03.jl
function priority(c::Char)::UInt
    islowercase(c) ? c - 'a' + 1 :
    isuppercase(c) ? c - 'A' + 27 :
    error("invalid character")
end

function find_error(str::AbstractString)::Char
    front, back = str[1:end÷2], str[end÷2+1:end]
    only(Set(front) ∩ Set(back))
end

function find_badge(block::AbstractString)::Char
    only(intersect(split(block)...))
end

function main()
    text = read("input/2022/input_03.txt", String)
    p1 = split(text) .|> find_error .|> priority |> sum
    println("Part 1: $p1")
    p2 = eachmatch(r"(\w+\n){3}", text) .|> (m -> m.match) .|> find_badge .|> priority |> sum
    println("Part 2: $p2")
end

main()

File: rust/2022/src/bin/day_03.rs
use std::collections::HashSet;

fn priority(c: u8) -> u8 {
    match c {
        b'a'..=b'z' => c - b'a' + 1,
        b'A'..=b'Z' => c - b'A' + 27,
        _ => panic!("input should be an ASCII letter"),
    }
}

fn find_duplicate(sack: &str) -> u8 {
    let (front, back) = sack.split_at(sack.len() / 2);
    let front_set: HashSet<_> = front.bytes().collect();
    let back_set: HashSet<_> = back.bytes().collect();
    *front_set.intersection(&back_set).next().unwrap()
}

fn find_badge(sacks: &[&str]) -> u8 {
    sacks
        .iter()
        .map(|sack| sack.bytes().collect::<HashSet<_>>())
        .reduce(|a, b| &a & &b)
        .unwrap()
        .into_iter()
        .next()
        .unwrap()
}

fn main() {
    let input = aoc::read_input("2022/input_03.txt");
    let sacks: Vec<_> = input.lines().collect();
    let p1 = sacks
        .iter()
        .map(|sack| priority(find_duplicate(sack)) as u32)
        .sum::<u32>();
    println!("Part 1: {p1}");
    let p2 = sacks
        .chunks(3)
        .map(|group| priority(find_badge(group)) as u32)
        .sum::<u32>();
    println!("Part 2: {p2}");
}
